library/pod-security-policy/selinux/src.rego

```diff
@@ -1,30 +1,33 @@
 package k8spspselinux
 
+# Disallow top level custom SELinux options
+violation[{"msg": msg, "details": {}}] {
+    has_field(input.review.object.spec.securityContext, "seLinuxOptions")
+    not input_seLinuxOptions_allowed(input.review.object.spec.securityContext.seLinuxOptions)
+    msg := sprintf("SELinux options is not allowed, pod: %v. Allowed options: %v", [input.review.object.metadata.name, input.parameters.allowedSELinuxOptions])
+}
+# Disallow container level custom SELinux options
 violation[{"msg": msg, "details": {}}] {
     c := input_security_context[_]
+    has_field(c.securityContext, "seLinuxOptions")
     not input_seLinuxOptions_allowed(c.securityContext.seLinuxOptions)
-    msg := sprintf("SELinux option is not allowed, pod: %v. Allowed options: %v", [input.review.object.metadata.name, input.parameters.allowedSELinuxOptions])
-}
-
-input_seLinuxOptions_allowed(options) {
-    input.parameters.allowedSELinuxOptions.level == options.level
+    msg := sprintf("SELinux options is not allowed, pod: %v, container %v. Allowed options: %v", [input.review.object.metadata.name, c.name, input.parameters.allowedSELinuxOptions])
 }
-
 input_seLinuxOptions_allowed(options) {
-    input.parameters.allowedSELinuxOptions.role == options.role
+    field_allowed(options, "level")
+    field_allowed(options, "role")
+    field_allowed(options, "type")
+    field_allowed(options, "user")
 }
 
-input_seLinuxOptions_allowed(options) {
-    input.parameters.allowedSELinuxOptions.type == options.type
+field_allowed(options, field) {
+    matching := {1 | input.parameters.allowedSELinuxOptions[_][field] == options[field]}
```

Comment:
Should we be making sure the "allowed" bit for level/role/type/user all come from the same list entry to avoid allowing a composite of all entries?

Reply:
Shoot, I thought a composite was desirable. Blocking composite makes sense though. I'll update so it blocks that